File: 2023/day2/cpp/main.cpp
#include <iostream>
#include <fstream>
#include <ranges>
#include <regex>

int extract_number(const std::string &str) {

    std::regex pattern("(?:Game (\\d*))|(?:(\\d*) green)|(?:(\\d*) red)|(?:(\\d*) blue)");
    std::smatch match;
    if (!std::regex_search(str, match, pattern)) {
        throw std::runtime_error("Could not match " + str);
    }
    std::cout << str << " " << match.size() << std::endl;
    for (int i = 0; i < match.size(); i++) {
        std::cout << i << ": " << match.str(i) << std::endl;
    }
    throw(std::runtime_error("Not implemented"));

    return 1;
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        std::cerr << "Usage: " << argv[0] << " <input file>" << std::endl;
        return 1;
    } 

    std::ifstream input(argv[1]);
    if (!input.is_open()) {
        std::cerr << "Could not open file: " << argv[1] << std::endl;
        return 1;
    }

    int sum = 0;
    while (!input.eof()) {
        std::string line;
        std::getline(input, line);
        sum += extract_number(line);
    }

    std::cout << "Sum: " << sum << std::endl;
    return 0;
}

File: 2023/day2/golang/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

func main() {
	if len(os.Args) < 2 {
		fmt.Printf("Usage: %s <input>\n", os.Args[0])
		return
	}

	input := os.Args[1]

	all, err := os.ReadFile(input)
	if err != nil {
		panic(err)
	}

	fmt.Println(string(all))
	file, err := os.Open(input)
	if err != nil {
		panic(err)
	}

	re, err := regexp.Compile("(?:Game (\\d*))|(?:(\\d*) red)|(?:(\\d*) green)|(?:(\\d*) blue)")
	if err != nil {
		panic(err)
	}

	reader := bufio.NewScanner(file)

	var (
		maxRed   int64 = 12
		maxGreen int64 = 13
		maxBlue  int64 = 14
	)

	maxValues := []int64{maxRed, maxGreen, maxBlue}

	var sumPart1 int64 = 0
	var sumPart2 int64 = 0

	for reader.Scan() {
		line := reader.Text()
		fmt.Println(line)
		submatches := re.FindAllStringSubmatch(line, -1)
		if len(submatches) == 0 {
			panic("no match")
		}
		game := submatches[0][1]
		gameInt, err := strconv.ParseInt(game, 10, 64)
		fmt.Println(gameInt)
		if err != nil {
			fmt.Println("game", game)
			panic(err)
		}
		valid := true
		minValues := []int64{-1, -1, -1}
		for _, submatch := range submatches[1:] {
			fmt.Println(submatch[2:])
			for colourIndex, val := range submatch[2:] {
				if val == "" {
					continue
				}
				valInt, err := strconv.ParseInt(val, 10, 64)
				if err != nil {
					panic(err)
				}
				if valInt > minValues[colourIndex] {
					minValues[colourIndex] = valInt
				}
				if valInt > maxValues[colourIndex] {
					fmt.Println(colourIndex, valInt, maxValues[colourIndex])
					fmt.Printf("Game %d, %s: %d > %d\n", gameInt, submatch, valInt, maxValues[colourIndex])
					valid = false
					continue
				}

			}
		}
		sumPart2 += minValues[0] * minValues[1] * minValues[2]

		if !valid {
			continue
		}

		sumPart1 += gameInt
	}
	fmt.Println(sumPart1)
	fmt.Println(sumPart2)

}
